### ISSUE TITLE
Error opening file /usr/share/ulauncher/preferences/dist/index.html: No such file or directory

### ISSUE THREAD

**Issue by voneAbides** [Ulauncher/Ulauncher#447]
### Steps to reproduce
1. Open Ulauncher

### Current behavior
When I open Ulauncher:
![Snímek z 2019-10-09 21-05-57](https://user-images.githubusercontent.com/44479267/66512626-9fb8c500-ead9-11e9-9683-486158c5e37e.png)

### Expected behavior
Ulauncher opens.

### Environment
**Ulauncher version**: 5.3.0-1
**Linux distro**: Fedora 30
**kernel**: 5.0.16-200.fc29.x86_64
**Desktop environment**: GNOME 3.32.2

### Logs
https://gist.github.com/voneAbides/9683e1102f18d90e44890490d5ab809c

**Comment by brpaz** [comment]
Got the same issue on Fedcora 30. It seems fixed after update to Ulauncher 5.4

**Comment by voneAbides** [comment]
Hello. I tried to update. Then I uninstalled and installed again. Still same Issue. Then Fedora 31 came out so I upgraded hoping that it could resolve the issue. Nothing worked.

**Comment by gornostal** [comment]
hi @voneAbides 
I've tried on Fedora 31 and preferences work fine for me.
Did you install ulauncher by downloading an RPM file from the website?

BTW, I've just updated RPM links on ulauncher.io. Now we have only one RPM build for Fedora 29, 30, and 31.

**Comment by voneAbides** [comment]
Hello @gornostal . I reinstalled with new package from Ulauncher page, but still only blank (white) screen.

**Comment by gornostal** [comment]
What command did you run to install the app?

Could you describe what directories and files you see in `/usr/share/ulauncher/preferences`?

**Comment by voneAbides** [comment]
@gornostal In /usr/share/ulauncher/preferences I see only dist folder.
I used `dnf install ulauncher.rpm`. Maybe long term solution will be add Ulauncher to Flathub.

**Comment by gornostal** [comment]
> I see only dist folder.

And I assume you don't have index.html in dist, right?
Are there any files in that directory?

**Comment by gornostal** [comment]
I just replied about Flathub in another issue.
It's a great thing, but I'd prefer keeping RPM and DEB up-to-date. At least for now.

**Comment by voneAbides** [comment]
Ok, sorry I forget that I uninstalled Ulauncher because it does not work. So I installed again. Same issue. I see these files and folders:
![Snímek z 2019-12-21 20-13-03](https://user-images.githubusercontent.com/44479267/71312720-8f548d00-242e-11ea-8a6d-b7f508ad09ad.png)
![Snímek z 2019-12-21 20-13-10](https://user-images.githubusercontent.com/44479267/71312722-91b6e700-242e-11ea-9b00-b974028094f8.png)

**Comment by voneAbides** [comment]
And preferences in gnome:
![Snímek z 2019-12-21 20-17-19](https://user-images.githubusercontent.com/44479267/71312762-30434800-242f-11ea-8069-c21c402cd227.png)

**Comment by gornostal** [comment]
Can you try running ulauncher from command line like this?
```
GDK_BACKEND=x11 ulauncher --no-extensions
```

Do you see any errors in the output?

**Comment by gornostal** [comment]
> Error opening file /usr/share/ulauncher/preferences/dist/index.html: No such file or directory
I suppose this error is gone now since you have that index.html file, right?

**Comment by voneAbides** [comment]
Yes, error is gone. This is output of `GDK_BACKEND=x11 ulauncher --no-extensions`
![Snímek z 2019-12-21 20-23-27](https://user-images.githubusercontent.com/44479267/71312818-f6bf0c80-242f-11ea-90da-67e44b3927a1.png)

**Comment by gornostal** [comment]
I see that ulauncher was running when you ran `GDK_BACKEND=x11 ulauncher --no-extensions`. Try to shut down the app first

**Comment by voneAbides** [comment]
Oh, sorry for that. This is the output:
![Snímek z 2019-12-21 20-32-49](https://user-images.githubusercontent.com/44479267/71312887-215d9500-2431-11ea-8189-f01e1bbc2eed.png)

**Comment by gornostal** [comment]
OK. that's better. And now could you open preferences while monitoring the output?

**Comment by voneAbides** [comment]
![Snímek z 2019-12-21 20-42-59](https://user-images.githubusercontent.com/44479267/71312983-8665ba80-2432-11ea-94e7-383315846f4f.png)

**Comment by gornostal** [comment]
anything unusual in the logs?

You can try one more thing:
If you run ulauncher with these arguments `GDK_BACKEND=x11 ulauncher --no-extensions --dev` you'll be able to right click on that white window and open Webkit inspector.
Then if you refresh the preferences by hitting Ctrl+R that could show an error in the Webkit inspector console.

Basically, it looks like preferences are not working due to some issue with webkit.

**Comment by voneAbides** [comment]
![Snímek z 2019-12-21 20-56-32](https://user-images.githubusercontent.com/44479267/71313110-8d8dc800-2434-11ea-96ac-7a81bbfa3ffa.png)

**Comment by voneAbides** [comment]
![Snímek z 2019-12-21 21-00-09](https://user-images.githubusercontent.com/44479267/71313142-f07f5f00-2434-11ea-9b08-837299ad0b8a.png)

**Comment by gornostal** [comment]
Thanks for debugging. I need to think how that could happen...

**Comment by voneAbides** [comment]
Ok, by the way. Ulauncher is great! :D

**Comment by gornostal** [comment]
@voneAbides the problem is not clear. I'm pretty sure it's caused by some customization of your system. Maybe you've installed some package that affects Ulauncher or changed settings in some way.

We can check one more thing in that debug console, but this time could you run select "Network" tab and take a screenshot?

For me it looks like this:
![Screenshot from 2019-12-22 21-45-54](https://user-images.githubusercontent.com/1202543/71326493-761e1000-2504-11ea-9c9b-c4fbfeb0c5da.png)

**Comment by voneAbides** [comment]
@gornostal Sorry, I was busy last days.
Network looks like this:
![Snímek z 2019-12-26 14-29-56](https://user-images.githubusercontent.com/44479267/71477818-60eaff00-27ec-11ea-83d7-77dfebdee151.png)

**Comment by Raqbit** [comment]
Hi there! I'm having the exact same issue.

For me, `index.html` is loaded from `file:///usr/share/ulauncher/preferences/dist/index.html#/preferences`, with 3 `/`'s.

Could it be that when webkit loads the `./static/..` content, something goes wrong and it tries to load from `file://null/` instead?

**Comment by gornostal** [comment]
@Raqbit you too use Fedora 31, right?
And by "having the same issue" you mean you see "Error opening file /usr/share/ulauncher/preferences/dist/index.html: No such file or directory" on a blank page?

`file:///usr/share/ulauncher/preferences/dist/index.html#/preferences` is the correct URL, btw.

Looking at @voneAbides's last screenshot I see that files are loaded correctly as there are no errors in the Network tab.

I need to reproduce it locally or in VBox somehow in order to fix the issue. So if you or @voneAbides can give me more details on how you configured your system that'd be great (What OS version, did you install any specific packages?, did you change any system settings that may affect this, etc)

**Comment by Raqbit** [comment]
@gornostal No, I'm using Ubuntu 19.10 with default (Yaru) theme and a Wayland session, I should have specified.

When I open the preferences, I get this white window:

![white window](https://user-images.githubusercontent.com/11144180/71644877-bda85700-2ccf-11ea-9bd0-97f2f334f831.png)

and then when I give any input it makes the window transparent:
![transparent window](https://user-images.githubusercontent.com/11144180/71644845-728e4400-2ccf-11ea-8643-747d2425b23e.png)

I must also note that the window is sometimes transparent from the get-go.

In devmode I find the following errors in the console tab:

![console tabs with errors](https://user-images.githubusercontent.com/11144180/71644897-4d4e0580-2cd0-11ea-9aa7-03d8533b5713.png)

and this is my Network tab:

![network tab without errors](https://user-images.githubusercontent.com/11144180/71644924-a322ad80-2cd0-11ea-823c-d43cc823a663.png)


Sometimes the window will half-render, but it won't be interactive as the javascript failed.
![image](https://user-images.githubusercontent.com/11144180/71644932-be8db880-2cd0-11ea-9ff1-4fa42a1b63e3.png)

**Comment by gornostal** [comment]
Thanks for the screenshots.
I don't know why that JS error occurs. I would need to reproduce it myself in order to figure out what's wrong.

The weird thing is I'm running PopOS based on Ubuntu 19.10. However I'm on Xorg.
I've tried Fedora 31 with Wayland in VBox, it works fine there.

I'll try Ubuntu 19.10 with Wayland in VBox and other distros too.

**Comment by voneAbides** [comment]
Now it works fine.

**Comment by gornostal** [comment]
ah I see the actual error now

I'm going to check the code. Hopefully I'll find why it can fail on that line...

**Comment by gornostal** [comment]
@voneAbides I don't know how to fix that particular error. I think it occurs because you may be running an old webkitgtk version. Try updating `webkit2gtk3` package.

Also, could you open this URL in Crhome please?
`file:///usr/share/ulauncher/preferences/dist/index.html?browser=1#/extensions`
It should open Preferences view with dummy content.

**Comment by voneAbides** [comment]
Ok. `webkit2gtk3` version is 2.26.2. It's up to date. URL looks like this:
![Snímek z 2020-01-21 15-55-43](https://user-images.githubusercontent.com/44479267/72815357-d1a7fe80-3c66-11ea-8449-ec506029e8bb.png)

Shortcuts content is weird.
![Snímek z 2020-01-21 15-58-51](https://user-images.githubusercontent.com/44479267/72815491-03b96080-3c67-11ea-8ac8-a24fd9cf3654.png)

**Comment by gornostal** [comment]
That's what is expected. It uses dummy content when there's word "browser"
in a URL.

So this means that in Chrome it works, but in webkitgkt it doesn't. And
it's only on your machine. I haven't been able to reproduce this issue
elsewhere.

**Comment by voneAbides** [comment]
Ok, thank you very much. I will try to fix it.

**Comment by Raqbit** [comment]
Fyi, I'm on webkit2gtk `2.26.3` (which is the latest version on ubuntu at the moment), issue still occurring.

Very weird issue, and because you are not able to reproduce it on your own system, maybe this is graphics driver related?

**Comment by gornostal** [comment]
@Raqbit I doubt it's graphics related, but if you can check whether it is, that would be great.
I have an integrated graphics card on all of my laptops.

**Comment by krmax44** [comment]
I have this problem as well. Is it because of Wayland?

**Comment by krmax44** [comment]
Okay, so by removing all source maps (all lines with `//# sourceMappingURL=...` from dist) the error messages go away. The window stays as is, though...

**Comment by algent-al** [comment]
Hi there! I am trying to package this application for Solus and have the exact same issue.
The file `index.html` isn't created here:
`/usr/share/ulauncher/preferences/dist/index.html` but it is created here:
`/usr/share/ulauncher/preferences/index.html`.

If I move `index.html` to `dist` folder, then I get only a `Loading...` page.

**Comment by friday** [comment]
This issue was almost certainly because of a bad fedora package release where the preferences web app weren't built.

As pointed out above, when packaging the app you either need to use the pre-built release tarball, or built it yourself with `./ul build-preferences` or `make prefs` (the makefile was just ported from v6)

A couple of the comments here were not directly related to the issue, but related to/fixed by ab5795f2e and e228aca00.
